pebble.go

```diff
@@ -0,0 +1,481 @@
+	return nil
+}
+
+// Delete implements Batch.
+func (b *pebbleDBBatch) Delete(key []byte) error {
+	//fmt.Println("pebbleDBBatch.Delete")
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+	if b.batch == nil {
+		return errBatchClosed
+	}
+	b.batch.Delete(key, nil)
+	return nil
+}
+
+// Write implements Batch.
+func (b *pebbleDBBatch) Write() error {
+	//fmt.Println("pebbleDBBatch.Write")
+	if b.batch == nil {
+		return errBatchClosed
+	}
+
+	wopts := pebble.NoSync
+	if isForceSync {
+		wopts = pebble.Sync
+	}
+	err := b.batch.Commit(wopts)
+	if err != nil {
+		return err
+	}
+	// Make sure batch cannot be used afterwards. Callers should still call Close(), for errors.
+
+	return b.Close()
+}
+
+// WriteSync implements Batch.
+func (b *pebbleDBBatch) WriteSync() error {
+	//fmt.Println("pebbleDBBatch.WriteSync")
+	if b.batch == nil {
+		return errBatchClosed
+	}
+	err := b.batch.Commit(pebble.Sync)
+	if err != nil {
+		return err
+	}
+	// Make sure batch cannot be used afterwards. Callers should still call Close(), for errors.
+	return b.Close()
+}
+
+// Close implements Batch.
+func (b *pebbleDBBatch) Close() error {
+	//fmt.Println("pebbleDBBatch.Close")
+	if b.batch != nil {
+		err := b.batch.Close()
+		if err != nil {
+			return err
+		}
+		b.batch = nil
+	}
+
+	return nil
+}
+
+type pebbleDBIterator struct {
+	source     *pebble.Iterator
+	start, end []byte
+	isReverse  bool
+	isInvalid  bool
+}
+
+var _ Iterator = (*pebbleDBIterator)(nil)
+
+func newPebbleDBIterator(source *pebble.Iterator, start, end []byte, isReverse bool) *pebbleDBIterator {
+	if isReverse {
+		if end == nil {
+			source.Last()
+		}
+	} else {
+		if start == nil {
+			source.First()
+		}
+	}
+	return &pebbleDBIterator{
+		source:    source,
+		start:     start,
+		end:       end,
+		isReverse: isReverse,
+		isInvalid: false,
+	}
+}
+
+// Domain implements Iterator.
+func (itr *pebbleDBIterator) Domain() ([]byte, []byte) {
+	//fmt.Println("pebbleDBIterator.Domain")
+	return itr.start, itr.end
+}
+
+// Valid implements Iterator.
+func (itr *pebbleDBIterator) Valid() bool {
+	//fmt.Println("pebbleDBIterator.Valid")
+	// Once invalid, forever invalid.
+	if itr.isInvalid {
+		return false
+	}
+
+	// If source has error, invalid.
+	if err := itr.source.Error(); err != nil {
+		itr.isInvalid = true
+
+		return false
+	}
+
+	// If source is invalid, invalid.
+	if !itr.source.Valid() {
+		itr.isInvalid = true
+
+		return false
+	}
+
+	// If key is end or past it, invalid.
+	start := itr.start
+	end := itr.end
+	key := itr.source.Key()
+	if itr.isReverse {
+		if start != nil && bytes.Compare(key, start) < 0 {
+			itr.isInvalid = true
+
+			return false
+		}
+	} else {
+		if end != nil && bytes.Compare(end, key) <= 0 {
+			itr.isInvalid = true
+
+			return false
+		}
+	}
+
+	// It's valid.
+	return true
+}
+
+// Key implements Iterator.
+func (itr *pebbleDBIterator) Key() []byte {
+	//fmt.Println("pebbleDBIterator.Key")
+	itr.assertIsValid()
+	return cp(itr.source.Key())
```

Comment:
Do we need to copy the byte array, can it just not be a 
```
return itr.source.Key()
```

Reply:
I think the second.

Reply:
need to copy, see https://github.com/cockroachdb/pebble/blob/v1.0.0/iterator.go#L2106